NEW VAT payment request and VAT payment are 2 different steps in workflow on VAT payment recording

atm-gauthier opened the pull request.

atm-gauthier committed:
NEW : add real payments on vat objects

atm-gauthier committed:
Merge branch '12.0_new_payments_on_vat' of github.com:atm-gauthier/dolibarr into NEW/add_real_payments_on_vat_objects

atm-gauthier committed:
FIX : VAT declaration list + migration script

atm-gauthier committed:
FIX : set paid button display test

atm-gauthier committed:
Merge branch '12.0_new_payments_on_vat' of github.com:atm-gauthier/dolibarr into NEW/add_real_payments_on_vat_objects

stickler-ci committed:
Fixing style errors.

eldy commented on htdocs/compta/bank/bankentries_list.php:
```diff
@@ -1209,10 +1210,11 @@
-					$paymentvatstatic->ref = $links[$key]['url_id'];
-					print ' '.$paymentvatstatic->getNomUrl(2);
-				} elseif ($links[$key]['type'] == 'payment_salary')
+                    print '<a href="'.DOL_URL_ROOT.'/compta/payment_vat/card.php?id='.$links[$key]['url_id'].'">';
```
Why removing a link with getNomUrl (recommended) with a hard coded value (must be avoid) ?

atm-gauthier replied:
@eldy i did new feature on 12 dolibarr version, it was a merge mistake, i just corrected that, is it ok for you ?

atm-gauthier committed:
FIX : getNomUrl()

atm-gauthier committed:
Merge branch '12.0_new_payments_on_vat' of github.com:atm-gauthier/dolibarr into NEW/add_real_payments_on_vat_objects

atm-gauthier committed:
Merge branch 'NEW/add_real_payments_on_vat_objects' of github.com:atm-gauthier/dolibarr into NEW/add_real_payments_on_vat_objects

atm-gauthier committed:
FIX : group by

atm-gauthier committed:
Merge branch '12.0_new_payments_on_vat' of github.com:atm-gauthier/dolibarr into NEW/add_real_payments_on_vat_objects

eldy commented on htdocs/install/mysql/migration/13.0.0-14.0.0.sql:
```diff
@@ -105,3 +105,36 @@ ALTER TABLE llx_propal ADD INDEX idx_propal_fk_warehouse(fk_warehouse);
+
+UPDATE llx_bank_url url INNER JOIN llx_tva tva ON tva.rowid = url.url_id SET type = 'vat', url.label = CONCAT('(', tva.label, ')') WHERE type = 'payment_vat';
+
+INSERT INTO `llx_bank_url` (fk_bank, url_id, url, label, type)
```
This generates error in CI.  Char ` is forbidden into sql because it is not portable.
We should not use such char because special chars are not allowed into table and fields in dolibarr rules.

atm-gauthier replied:
Hi @eldy, no problem i removed them

atm-gauthier replied:
@eldy is everything ok for you ?

atm-gauthier committed:
FIX : remove backquotes

atm-gauthier committed:
FIX : double quotes on sql file

atm-gauthier committed:
FIX : sql db escape missing

atm-gauthier committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into NEW/add_real_payments_on_vat_objects

eldy commented on htdocs/install/mysql/migration/13.0.0-14.0.0.sql:
```diff
@@ -121,3 +121,36 @@ ALTER TABLE llx_societe ADD INDEX idx_societe_warehouse(fk_warehouse);
+WHERE type = 'vat';
+
+ALTER TABLE llx_tva DROP COLUMN fk_bank;
+UPDATE llx_tva SET paye = 1;
```
There is an important rule in migration process. We must be able to run/rerun the upgrade process with no risk to loose data. Here, if we do that, we will lose the correct status.
The solution I see is 
- create field in migration with default value to 1 (so old field will have value 1)
- update the default value of field to 0

eldy commented on htdocs/install/mysql/migration/13.0.0-14.0.0.sql:
```diff
@@ -121,3 +121,36 @@ ALTER TABLE llx_societe ADD INDEX idx_societe_warehouse(fk_warehouse);
+INSERT INTO llx_bank_url (fk_bank, url_id, url, label, type)
+SELECT b.fk_bank, ptva.rowid, REPLACE(b.url, 'tva/card.php', 'payment_vat/card.php'), '(paiement)', 'payment_vat'
+FROM llx_bank_url b INNER JOIN llx_tva tva ON (tva.fk_bank = b.fk_bank) INNER JOIN llx_paiementtva ptva on (ptva.fk_bank = b.fk_bank)
+WHERE type = 'vat';
```
What if we rerun the upgrade process twice ? What happen into llx_paiementtva and llx_bank_url ?
If not, just remove this part. It is not a problem if we don't have data of a "new feature" for "past record".

eldy commented on htdocs/install/mysql/migration/13.0.0-14.0.0.sql:
```diff
@@ -121,3 +121,36 @@ ALTER TABLE llx_societe ADD INDEX idx_societe_warehouse(fk_warehouse);
+ALTER TABLE llx_tva ADD paye smallint default 0 NOT NULL;
+ALTER TABLE llx_tva ADD fk_account integer;
+
+INSERT INTO llx_paiementtva (fk_tva, datec, datep, amount, fk_typepaiement, num_paiement, note, fk_bank, fk_user_creat, fk_user_modif)
```
For new table, we must try to use english version. Can you use instead
llx_payment_vat
to match llx_payment_salary or llx_payment _various

atm-gauthier committed:
FIX : table name for VAT payments must be llx_payment_vat

atm-gauthier committed:
Merge branch '12.0_new_payments_on_vat' of github.com:atm-gauthier/dolibarr into NEW/add_real_payments_on_vat_objects

atm-gauthier committed:
FIX : migration script

atm-gauthier committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into NEW/add_real_payments_on_vat_objects

atm-gauthier commented:
Hi @eldy
- I renamed llx_paiementtva into llx_payment_vat
- I used a default value of 1 for "paye" and set it to 0 at the end of the script
- We won't meet no error if we rerun migration script for these reasons : 
    - INSERT query on llx_payment_vat will work only one time because fk_bank field of llx_tva is deleted at the end of the script
    - INSERT query on llx_bank_url will work only one time because fk_bank field of llx_tva is deleted at the end of the script
    - UPDATE query on llx_bank_url will work only one time because there is a constraint which not allows to INSERT same url for same object id
    
Is everything ok for you now ?

eldy committed:
Several fixes in migration

Fix we must not drop a field during migration. This must be done 2 or 3 versions later.
Fix syntax of sql

atm-gauthier committed:
Merge branch 'develop' of github.com:Dolibarr/dolibarr into NEW/add_real_payments_on_vat_objects
